subprojects/core-api/src/main/java/org/gradle/api/provider/ListProperty.java

```diff
@@ -61,4 +64,47 @@ public interface ListProperty<T> extends Provider<List<T>>, HasMultipleValues<T>
      */
     @Override
     ListProperty<T> convention(Provider<? extends Iterable<? extends T>> provider);
+
+    /**
+     * Applies an eager transformation to the current value of the property "in place", without explicitly obtaining it.
+     * The provided transformer is applied to the provider of the current value, and the returned provider is used as a new value.
+     * The provider of the value can be used to derive the new value, but doesn't have to.
+     * Returning null from the transformer unsets the property.
+     * For example, the current value of a string list property can be reversed:
+     * <pre class='autoTested'>
+     *     def property = objects.listProperty(String).value(["a", "b"])
+     *
+     *     property.update { it.map { value -&gt; value.reverse() } }
+     *
+     *     println(property.get()) // ["b", "a"]
+     * </pre>
+     * Note that simply writing {@code property.set(property.map { ... } } doesn't work and will cause an exception because of a circular reference evaluation at runtime.
+     * <p>
+     * <b>Further changes to the value of the property, such as calls to {@link #set(Iterable)}, are not transformed, and override the update instead</b>.
+     * Because of this, this method inherently depends on the order of property changes, and therefore must be used sparingly.
+     * <p>
+     * If the value of the property is specified via a provider, then the current value provider tracks that provider.
+     * For example, changes to the upstream property are visible:
+     * <pre class='autoTested'>
+     *     def upstream = objects.listProperty(String).value(["a", "b"])
+     *     def property = objects.listProperty(String).value(upstream)
+     *
+     *     property.update { it.map { value -&gt; value.reverse() } }
+     *     upstream.set(["c", "d"])
+     *
+     *     println(property.get()) // ["d", "c"]
+     * </pre>
+     * The provided transformation runs <b>eagerly</b>, so it can capture any objects without introducing memory leaks and breaking configuration cache.
```

Comment:
```suggestion
     * The provided transformation runs <b>eagerly</b>, so it can capture any objects without introducing memory leaks and without breaking configuration caching.
```